proposals/Simple-programs.md

```diff
@@ -0,0 +1,169 @@
+
+``` c#
+static class Program
+{
+    static async Task Main()
+    {
+        // statements
+    }
+}
+```
+
+See also https://github.com/dotnet/csharplang/issues/3117.
+
+## Motivation
+[motivation]: #motivation
+
+There's a certain amount of boilerplate surrounding even the simplest of programs,
+because of the need for an explicit `Main` method. This seems to get in the way of
+language learning and program clarity. The primary goal of the feature therefore is
+to allow C# programs without unnecessary boilerplate around them, for the sake of
+learners and the clarity of code.
+
+## Detailed design
+[design]: #detailed-design
+
+### Syntax
+
+The only additional syntax is allowing a sequence of *statement*s in a compilation unit,
+just before the *namespace_member_declaration*s:
+
+``` antlr
+compilation_unit
+    : extern_alias_directive* using_directive* global_attributes? statement* namespace_member_declaration*
+    ;
+```
+
+In all but one *compilation_unit* the *statement*s must all be local function declarations. 
+
+Example:
+
+``` c#
+// File 1 - any statements
+if (args.Length == 0
+    || !int.TryParse(args[0], out int n)
+    || n < 0) return;
+Console.WriteLine(Fib(n).curr);
+
+// File 2 - only local functions
+(int curr, int prev) Fib(int i)
+{
+    if (i == 0) return (1, 0);
+    var (curr, prev) = Fib(i - 1);
+    return (curr + prev, curr);
+}
+```
+
+### Semantics
+
+If any top-level statements are present in any compilation unit of the program, the meaning is as if
+they were combined in the block body of a `Main` method of a `Program` class in the global namespace,
+as follows:
+
+``` c#
+static class Program
+{
+    static async Task Main()
+    {
+        // File 1 statements
+        // File 2 local functions
+        // ...
+    }
+}
+```
+
+Note that the names "Program" and "Main" are used only for illustrations purposes, actual names used by
+compiler are implementation dependent and neither the type, nor the method can be referenced by name from
+source code.
+
+The method is designated as the entry point of the program. Explicitly declared methods that by convention 
+could be considered as an entry point candidates are ignored. A warning is reported when that happens. It is
+an error to specify `-main:<type>` compiler switch.
+
+If any one compilation unit has statements other than local function declarations, statements from that
+compilation unit occur first. This causes it to be legal for local functions in one file to reference
+local variables in another. The order of statement contributions (which would all be local functions)
+from other compilation units is undefined.
+
+If `await` expressions and other async operations are omitted, no warning is produced. Instead the
+signature of the generated entry point method is equivalent to 
+``` c#
+    static void Main()
+```
+
+The example above would yield the following `$Main` method declaration:
+
+``` c#
+static class $Program
```

Comment:
Do you plan to actually emit a hidden type for this, or is the Main going to emitted as a global function in this case?

Reply:
The current implementation emits a hidden type: https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQGUCWBbADgGwgAUEB7AcwSm2QBoYRN8AfAAQCYBGAWAChXOATgAUAIgAWEfPlKiAlAG4+AAlXKgA

Reply:
> Do you plan to actually emit a hidden type for this, or is the Main going to emitted as a global function in this case?

The current plan is to synthesize a type with a method inside it.

Reply:
Is there a reason why not emit it as global function? The global functions were kind of meant for this. Just wondering.

Reply:
> Is there a reason why not emit it as global function? The global functions were kind of meant for this. Just wondering.

I do not believe there is a good reason to complicate things. As it stands right now, compiler never emits global functions, not to mention global functions with user code. Our pipeline from beginning to the end assumes there is a type and that works just fine for this scenario.